providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+	// Example: X-BMCLIB-Signature becomes X-BMCLIB-Signature-256
+	// When set to true, a header will be added for each algorithm. Example: X-BMCLIB-Signature-256 and X-BMCLIB-Signature-512
+	AppendAlgoToHeaderDisabled bool
+	// IncludedPayloadHeaders are headers whose values will be included in the signature payload. Example: X-BMCLIB-My-Custom-Header
+	// All headers will be deduplicated.
+	IncludedPayloadHeaders []string
+}
+
+type HMACOpts struct {
+	// Hashes is a map of algorithms to a slice of hash.Hash (these are the hashed secrets). The slice is used to support multiple secrets.
+	Hashes map[Algorithm][]hash.Hash
+	// PrefixSigDisabled determines whether the algorithm will be prefixed to the signature. Example: sha256=abc123
+	PrefixSigDisabled bool
+	// Secrets are a map of algorithms to secrets used for signing.
+	Secrets Secrets
+}
+
+type Experimental struct {
+	// CustomRequestPayload must be in json.
+	CustomRequestPayload []byte
+	// DotPath is the path to where the bmclib RequestPayload{} will be embedded. For example: object.data.body
+	DotPath string
+}
+
+// New returns a new Config containing all the defaults for the rpc provider.
+func New(consumerURL string, host string, secrets Secrets) *Config {
+	// defaults
+	c := &Config{
+		Host:        host,
+		ConsumerURL: consumerURL,
+		Logger:      logr.Discard(),
+		Opts: Opts{
+			Request: RequestOpts{
+				Client:          http.DefaultClient,
+				HTTPContentType: contentType,
+				HTTPMethod:      http.MethodPost,
+				TimestampFormat: time.RFC3339,
+				TimestampHeader: timestampHeader,
+			},
+			Signature: SignatureOpts{
+				HeaderName:             signatureHeader,
+				IncludedPayloadHeaders: []string{},
+			},
+			HMAC: HMACOpts{
+				Hashes:  map[Algorithm][]hash.Hash{},
+				Secrets: secrets,
+			},
+			Experimental: Experimental{},
+		},
+	}
+
+	if len(secrets) > 0 {
+		c.Opts.HMAC.Hashes = CreateHashes(secrets)
+	}
+
+	return c
+}
+
+// Name returns the name of this rpc provider.
+// Implements bmc.Provider interface
+func (c *Config) Name() string {
+	return ProviderName
+}
+
+// Open a connection to the rpc consumer.
+// For the rpc provider, Open means validating the Config and
+// that communication with the rpc consumer can be established.
+func (c *Config) Open(ctx context.Context) error {
+	// 1. validate consumerURL is a properly formatted URL.
+	// 2. validate that we can communicate with the rpc consumer.
+	u, err := url.Parse(c.ConsumerURL)
+	if err != nil {
+		return err
+	}
+	c.listenerURL = u
+	testReq, err := http.NewRequestWithContext(ctx, c.Opts.Request.HTTPMethod, c.listenerURL.String(), nil)
+	if err != nil {
+		return err
+	}
+	// test that we can communicate with the rpc consumer.
+	// and that it responses with the spec contract (Response{}).
+	if _, err := c.Opts.Request.Client.Do(testReq); err != nil { //nolint:bodyclose // not reading the body
```

Comment:
would be ideal the whole body and close - so the connection can be reused (no?)

https://pkg.go.dev/net/http#Client.Do
```
If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.
```